README

Zoovy/AnyCommerce MVC JQuery Developer Framework

Rich Internet Application, Model View Controller created using jQuery for creating standalone or 
pluggable shopping applications that connect to the Zoovy e-Commerce Platform.

WIKI Documentation:
	https://github.com/zoovy/AnyCommerce-Development/wiki

Official Repository:
	https://github.com/zoovy/AnyCommerce-Development	

API Calls Input/Output Reference:
	http://www.zoovy.com/webdoc/index.cgi?VERB=DOC&DOCID=51609

Examples are stored in examples sub-directory.
Extensions are stored in the extensions sub-directory.



changelog.txt
#
# This file will be updated with each release (duh)
#

Any time you are upgrading to a new version of the MVC, you should create a local backup of the site, so that you can easily restore or refer back as necessary.

2012.31 >  
See git repository for changes. Individual branches contain change logs.

2012.31 changes

* controller.js
 -> updated util.getSupplementalPaymentInputs (need to test PO and ECHECK in C/O)

* store_crm.js
 -> call added for buyerAddressAddUpdate

* quickstart.js
 -> callbacks for buyer to update address.
 -> util for jeditable.
 -> Fixed bug where uri params got nuked on landing page (were in memory, but not on URI which may be necessary for third party scripts)
 -> moved the createTemplateFunctions to top of start. better for third party plugins that use templateFunctions.

* store_product.js
 -> fixed issue where prodlist vars were getting dropped when pagination was used (effected withReviews setting)


2012.28 changes

* model.js
 -> added feature to allow for loading .json (flat files) of product data. Useful for native apps.
 -> Changed the # of cmds in a pipelined request to a var instead of a fixed #. Will later allow a debug mode to excersize more control.
 -> immutableQ no longer aborts mutable. Now, for a Q or request to get aborted, the app must do it.
 -> added a supported param in _tag for 'forceError'. set to 1 and the 'onError' function (if a callback is specified) will get reached instead of onSuccess (even if call is successful).
 -> started adding some support for planned debugger tool.

* controller.js
 -> youtube renderFormat now supports a height/width param (set in data-bind) for more control over appearance. default height/width set.

* checkout_nice and checkout_passive
 -> updated so that when the form validation and order creation is taking place, a loading modal appears w/ status updates.

* store_checkout.js
 -> fixed major bug where payment info (cc) was not being sent on buyerOrderCreate (bug introduced in 201225)

* store_cart.js
 -> added new call 'cartItemsInventoryVerify' to allow a check of the cart (prior to checkout) for inventory availability.

* store_product.js
 -> fixed bug in addToCart error that cause messaging to not consistently appear.

* store_prodlist.js
 -> finished adding support for reviews in a prodlist.

* store_search.js
 -> deprecated legacy search. keyword now uses a native elastic search.

* quickstart.html
A big change to the way the MVC is loaded.  Instead of directly including all the .js files, only two are included (jquery and loadscript). There is a new appPreView ID that contains a loader, some text, and the merchants logo (which needs to be hard coded and should be a fairly small file). The html5 progress bar is updated as the inlcudes are loaded.  

There is a new file, init.js which is the first file included via loadscript.  It contains all the references to ALL the other files that need to be included, including any third party scripts. 

 -> productListTemplateChildren to be clickable.
 -> price/msrp/savings updated to use new render formats (see quickstart.js update)

* quickstart.js
 -> added new function for changing body/html cursor.
 -> added onInit and onComplete functions to change cursor to 'wait' and 'auto' respectively.
 -> new render formats for priceRetailSavingsPercentage and priceRetailSavingsDifference

* qs_styles
 -> adjusted loadingBG classes to reference new loading/wait animated gifs



2012.25 changes

Added a 'snippets.html' file to the examples directory. Contains some useful snippets for elements such as search and newsletter. More will be added...

model.js
 -> Bug fix in extensions loader. Will now check to make sure all dependencies templates have loaded before executing the init callback.

* controller.renderFunctions
 -> changed the way pre/post text work.  They are no longer added to the value before it is passed into the format function. Instead, they're handled afterwards. This means that bindData.cleanValue is no longer necessary (value is always 'clean' now) and has been removed. pre/post text should be used for text only (does not support html).
 -> support for before, after and wrap have been added globally. These should be used for adding any html.
 -> added myControl.vars.protocol which will = https: (if secure) or http: (if non-secure or local). This may change as native app support progresses.

Checkouts
 -> panels.chkoutPayOptionsFieldset had the supplemental payment information removed. It was added as a function in the controller so it could be more global (it'll be shared in checkout, admin orders and customer order updates.

* store_prodlist
 -> removed the 'getProductDataForLaterUse' function. Use getProductDataForList instead (they were very similar functions). 

* store_crm
 -> started support for buyerOrderMacro (ability for a buyer to update their order)

* store_cart
 -> started amazon payments integration

* store_product
 -> removed util.getXsellForPID (not necessary w/ buildQueriesFromTemplate code)
 
* quickstart
 -> change to addClass renderFormat (required because of changes to pre/posttext)
 -> some changes to reviews, including new templates and bindData's (productTemplate). new reviews img with all 11 ratings.
 -> broke changeNavTo out into separate functions, each called by showContent as needed:
	showSearch
	showCompany
	showCustomer
	This change was necessary because of a bug where landing on a company page resulted in an error because the profile data hadn't been loaded yet.  This is cleaner too.
 -> as part of removing changeNavTo, added a function for util.thisArticleRequiresLogin(P) (article is what a 'show' value is internally referenced as (ex: return or myaccount is an article). will return true or false based on whether P.show is a secure article (ex: show=orders requires login).

admin_init.js
var app = app || {vars:{},u:{}}; //make sure app exists.
app.rq = app.rq || []; //ensure array is defined. rq = resource queue.




app.rq.push(['extension',1,'store_prodlist','extensions/store_prodlist.js']);
app.rq.push(['extension',1,'store_navcats','extensions/store_navcats.js']);
app.rq.push(['extension',1,'store_search','extensions/store_search.js']);
app.rq.push(['extension',1,'store_product','extensions/store_product.js']);
app.rq.push(['extension',1,'admin_prodEdit','extensions/admin/product_editor.js']);
app.rq.push(['extension',0,'admin','extensions/admin/extension.js','initUserInterface']);


app.rq.push(['script',0,app.vars.baseURL+'model.js']); //'validator':function(){return (typeof zoovyModel == 'function') ? true : false;}}
app.rq.push(['script',0,app.vars.baseURL+'includes.js']); //','validator':function(){return (typeof handlePogs == 'function') ? true : false;}})
app.rq.push(['script',0,app.vars.baseURL+'controller.js']);
app.rq.push(['script',1,app.vars.baseURL+'jeditable.js']); //used for making text editable (customer address). non-essential. loaded late.


app.rq.push(['script',1,'https://www.zoovy.com/biz/ajax/showLoading/js/jquery.showLoading.min.js']);
app.rq.push(['script',1,'https://www.zoovy.com/biz/ajax/flexigrid-1.1/js/flexigrid.js']);
app.rq.push(['script',1,'https://www.zoovy.com/biz/ajax/jquery.qtip-1.0.0-rc3/jquery.qtip-1.0.0-rc3.min.js']);
app.rq.push(['script',1,'https://www.zoovy.com/biz/ajax/jquery.contextMenu/jquery.contextMenu.js']);
app.rq.push(['script',1,'https://www.zoovy.com/biz/ajax/jquery.popup/jquery.popupWindow.js']);


//group any third party files together (regardless of pass) to make troubleshooting easier.
app.rq.push(['script',0,(document.location.protocol == 'https:' ? 'https:' : 'http:')+'//ajax.googleapis.com/ajax/libs/jqueryui/1.9.0/jquery-ui.js']);



/*
This function is overwritten once the controller is instantiated. 
Having a placeholder allows us to always reference the same messaging function, but not impede load time with a bulky error function.
*/
app.u.throwMessage = function(m)	{
	alert(m); 
	}

app.u.howManyPassZeroResourcesAreLoaded = function(debug)	{
	var L = app.vars.rq.length;
	var r = 0; //what is returned. total # of scripts that have finished loading.
	for(var i = 0; i < L; i++)	{
		if(app.vars.rq[i][app.vars.rq[i].length - 1] === true)	{
			r++;
			}
		if(debug)	{app.u.dump(" -> "+i+": "+app.vars.rq[i][2]+": "+app.vars.rq[i][app.vars.rq[i].length -1]);}
		}
	return r;
	}


//gets executed once controller.js is loaded.
//check dependencies and make sure all other .js files are done, then init controller.
//function will get re-executed if not all the scripts in app.vars.scripts pass 1 are done loading.
//the 'attempts' var is incremented each time the function is executed.

app.u.initMVC = function(attempts){
//	app.u.dump("app.u.initMVC activated ["+attempts+"]");
	var resourcesLoaded = app.u.howManyPassZeroResourcesAreLoaded();
	if(resourcesLoaded == app.vars.rq.length)	{
//instantiate controller. handles all logic and communication between model and view.
//passing in app will extend app so all previously declared functions will exist in addition to all the built in functions.
//tmp is a throw away variable. app is what should be used as is referenced within the mvc.
		app.vars.rq = null; //to get here, all these resources have been loaded. nuke record to keep DOM clean and avoid any duplication. note this is NOT app.rq
		var tmp = new zController(app);
//instantiate wiki parser.
		myCreole = new Parse.Simple.Creole();
		}
	else if(attempts > 50)	{
		app.u.dump("WARNING! something went wrong in init.js");
		//this is 10 seconds of trying. something isn't going well.
		$('.appMessaging').empty().append("<h2>Uh Oh. Something seems to have gone wrong. </h2><p>Several attempts were made to load the store but some necessary files were not found or could not load. We apologize for the inconvenience. Please try 'refresh' and see if that helps.<br><b>If the error persists, please contact the site administrator</b><br> - dev: see console.</p>");
		app.u.howManyPassZeroResourcesAreLoaded(true);
		}
	else	{
		setTimeout("app.u.initMVC("+(attempts+1)+")",250);
		}

	}



//don't execute script till both jquery AND the dom are ready.
$(document).ready(function(){
	app.u.handleRQ(0); //will start loading resources and eventually init the app.
	});








loadscript.js
var app = app || {vars:{},u:{}}; //make sure app exists.

//http://www.nczonline.net/blog/2009/07/28/the-best-way-to-load-external-javascript/
app.u.loadScript = function(url, callback, params){
//	app.u.dump("load script: "+url);
    var script = document.createElement("script");
    script.type = "text/javascript";
    if (script.readyState){  //IE
        script.onreadystatechange = function(){
            if (script.readyState == "loaded" || script.readyState == "complete"){
                script.onreadystatechange = null;
				if(typeof callback == 'function')	{callback(params);}
				}
			};
		}
	else {
		if(typeof callback == 'function')	{
			script.onload = function(){callback(params)}
			}
    	}
    script.src = url;
    document.getElementsByTagName("head")[0].appendChild(script);
	}



/*
Will load all scripts and extenstions with pass = 0.
pass with any other value (including blank,null, undefined, etc) will get loaded later.
this function is overwritten once the myRIA callback occurs with a similar but more powerful function (ignores pass, supports css).
app.rq.push() = app.u.handleResourceQ so whatever the values in push() are get executed immediately.
*/

app.u.handleRQ = function()	{
	app.u.dump("BEGIN app.u.handleRQ");
	var numIncludes = 0; //what is returned. The total number of includes for this pass.
	var L = app.rq.length - 1;

	app.vars.extensions = app.vars.extensions || []; //ensure array is defined.
	app.vars.rq = new Array(); //to avoid any duplication, as iteration occurs, items are moved from app.rq into this tmp array. 

//the callback added to the loadScript on type 'script' sets the last value of the resource array to true.
//another script will go through this array and make sure all values are true for validation. That script will execute the callback (once all scripts are loaded).
	var callback = function(index){
		app.vars.rq[index][app.vars.rq[index].length - 1] = true; //last index in array is for 'is loaded'. set to false in loop below.
		}
	
	for(var i = L; i >= 0; i--)	{
//		app.u.dump("app.rq["+i+"][0]: "+app.rq[i][0]+" pass: "+app.rq[i][1]);
		if(app.rq[i][0] == 'script' && app.rq[i][1] === 0)	{
//			app.u.dump(" -> load it!");
			numIncludes++;
			app.rq[i][app.rq[i].length] = false; //will get set to true when script loads as part of callback.
			app.vars.rq.push(app.rq[i]); //add to pass zero rq.
			app.u.loadScript(app.rq[i][2],callback,(app.vars.rq.length - 1));
			app.rq.splice(i, 1); //remove from new array to avoid dupes.
			}
		else if(app.rq[i][0] == 'extension' && app.rq[i][1] === 0)	{
			numIncludes++;
			app.vars.extensions.push({"namespace":app.rq[i][2],"filename":app.rq[i][3],"callback":app.rq[i][4]}); //add to extension Q.
			app.rq[i][app.rq[i].length] = false; //will get set to true when script loads as part of callback.
			app.vars.rq.push(app.rq[i]); //add to pass zero rq.

//on pass 0, no callbacks added to extensions because the model already has a function for checking if extensions are loaded.
// adding these extensions to the extensions array is necessary for this checker to work.
			app.u.loadScript(app.rq[i][3],callback,(app.vars.rq.length - 1));
			app.rq.splice(i, 1); //remove from old array to avoid dupes.
			}
		else	{
//currently, this function is intended for pass 0 only, so if an item isn't pass 0,do nothing with it.
			}
		}
//	app.u.dump("numIncludes: "+numIncludes);
	app.u.initMVC(0);
	return numIncludes;

	}



/*
this function gets overwritten when the control object is instantiated.
keep this small and light.
*/
app.u.dump = function(msg)	{
//if the console isn't open, an error occurs, so check to make sure it's defined. If not, do nothing.
	if(typeof console != 'undefined')	{console.log(msg);}
	} //dump

_blueCollarDepot_darrells_init.js
var app = app || {vars:{},u:{}}; //make sure app exists.
app.rq = app.rq || []; //ensure array is defined. rq = resource queue.




app.rq.push(['extension',0,'convertSessionToOrder','extensions/checkout_passive/extension.js']);
// app.rq.push(['extension',0,'convertSessionToOrder','extensions/checkout_nice/extension.js']);
app.rq.push(['extension',0,'store_checkout','extensions/store_checkout.js']);
app.rq.push(['extension',0,'store_prodlist','extensions/store_prodlist.js']);
app.rq.push(['extension',0,'store_navcats','extensions/store_navcats.js']);
app.rq.push(['extension',0,'store_search','extensions/store_search.js']);
app.rq.push(['extension',0,'store_product','extensions/store_product.js']);
app.rq.push(['extension',0,'store_cart','extensions/store_cart.js']);
app.rq.push(['extension',0,'store_crm','extensions/store_crm.js']);
app.rq.push(['extension',0,'myRIA','quickstart.js','startMyProgram']);

app.rq.push(['extension',1,'analytics_google','extensions/analytics_google.js','addTriggers']);
//app.rq.push(['extension',1,'bonding_buysafe','extensions/bonding_buysafe.js','addTriggers']);
//app.rq.push(['extension',1,'powerReviews','extensions/reviews_powerreviews.js','startExtension']);
//app.rq.push(['extension',0,'magicToolBox','extensions/imaging_magictoolbox.js','startExtension']); // (not working yet - ticket in to MTB)



//add tabs to product data.
app.rq.push(['templateFunction','productTemplate','onCompletes',function(P) {
	$( ".tabbedProductContent",$('#productTemplate_'+app.u.makeSafeHTMLId(P.pid))).tabs();
	}]);

app.rq.push(['script',0,(document.location.protocol == 'file:') ? app.vars.httpURL+'jquery/config.js' : app.vars.baseURL+'jquery/config.js']); //The config.js is dynamically generated.
app.rq.push(['script',0,app.vars.baseURL+'model.js']); //'validator':function(){return (typeof zoovyModel == 'function') ? true : false;}}
app.rq.push(['script',0,app.vars.baseURL+'includes.js']); //','validator':function(){return (typeof handlePogs == 'function') ? true : false;}})
app.rq.push(['script',1,app.vars.baseURL+'jeditable.js']); //used for making text editable (customer address). non-essential. loaded late.
app.rq.push(['script',0,app.vars.baseURL+'controller.js']);

//sample of an onDeparts. executed any time a user leaves this page/template type.
app.rq.push(['templateFunction','homepageTemplate','onDeparts',function(P) {app.u.dump("just left the homepage")}]);


//group any third party files together (regardless of pass) to make troubleshooting easier.
app.rq.push(['script',0,(document.location.protocol == 'https:' ? 'https:' : 'http:')+'//ajax.googleapis.com/ajax/libs/jqueryui/1.9.0/jquery-ui.js']);


/*
This function is overwritten once the controller is instantiated. 
Having a placeholder allows us to always reference the same messaging function, but not impede load time with a bulky error function.
*/
app.u.throwMessage = function(m)	{
	alert(m); 
	}

app.u.howManyPassZeroResourcesAreLoaded = function(debug)	{
	var L = app.vars.rq.length;
	var r = 0; //what is returned. total # of scripts that have finished loading.
	for(var i = 0; i < L; i++)	{
		if(app.vars.rq[i][app.vars.rq[i].length - 1] === true)	{
			r++;
			}
		if(debug)	{app.u.dump(" -> "+i+": "+app.vars.rq[i][2]+": "+app.vars.rq[i][app.vars.rq[i].length -1]);}
		}
	return r;
	}


//gets executed once controller.js is loaded.
//check dependencies and make sure all other .js files are done, then init controller.
//function will get re-executed if not all the scripts in app.vars.scripts pass 1 are done loading.
//the 'attempts' var is incremented each time the function is executed.

app.u.initMVC = function(attempts){
//	app.u.dump("app.u.initMVC activated ["+attempts+"]");
	var includesAreDone = true;

//what percentage of completion a single include represents (if 10 includes, each is 10%).
	var percentPerInclude = Math.round((100 / app.vars.rq.length));  
	var resourcesLoaded = app.u.howManyPassZeroResourcesAreLoaded();
	var percentComplete = resourcesLoaded * percentPerInclude; //used to sum how many includes have successfully loaded.

	$('#appPreViewProgressBar').val(percentComplete);
	$('#appPreViewProgressText').empty().append(percentComplete+"% Complete");

	if(resourcesLoaded == app.vars.rq.length)	{
//instantiate controller. handles all logic and communication between model and view.
//passing in app will extend app so all previously declared functions will exist in addition to all the built in functions.
//tmp is a throw away variable. app is what should be used as is referenced within the mvc.
		app.vars.rq = null; //to get here, all these resources have been loaded. nuke record to keep DOM clean and avoid any duplication.
		var tmp = new zController(app);
//instantiate wiki parser.
		myCreole = new Parse.Simple.Creole();
		}
	else if(attempts > 50)	{
		app.u.dump("WARNING! something went wrong in init.js");
		//this is 10 seconds of trying. something isn't going well.
		$('#appPreView').empty().append("<h2>Uh Oh. Something seems to have gone wrong. </h2><p>Several attempts were made to load the store but some necessary files were not found or could not load. We apologize for the inconvenience. Please try 'refresh' and see if that helps.<br><b>If the error persists, please contact the site administrator</b><br> - dev: see console.</p>");
		app.u.howManyPassZeroResourcesAreLoaded(true);
		}
	else	{
		setTimeout("app.u.initMVC("+(attempts+1)+")",250);
		}

	}



//Any code that needs to be executed after the app init has occured can go here.
//will pass in the page info object. (pageType, templateID, pid/navcat/show and more)
app.u.appInitComplete = function(P)	{
	app.u.dump("Executing myAppIsLoaded code...");
	}




//don't execute script till both jquery AND the dom are ready.
$(document).ready(function(){
	app.u.handleRQ(0)
	});








init.js
var app = app || {vars:{},u:{}}; //make sure app exists.
app.rq = app.rq || []; //ensure array is defined. rq = resource queue.




app.rq.push(['extension',0,'convertSessionToOrder','extensions/checkout_passive/extension.js']);
//app.rq.push(['extension',0,'convertSessionToOrder','extensions/checkout_nice/extension.js']);
app.rq.push(['extension',0,'store_checkout','extensions/store_checkout.js']);
app.rq.push(['extension',0,'store_prodlist','extensions/store_prodlist.js']);
app.rq.push(['extension',0,'store_navcats','extensions/store_navcats.js']);
app.rq.push(['extension',0,'store_search','extensions/store_search.js']);
app.rq.push(['extension',0,'store_product','extensions/store_product.js']);
app.rq.push(['extension',0,'store_cart','extensions/store_cart.js']);
app.rq.push(['extension',0,'store_crm','extensions/store_crm.js']);
app.rq.push(['extension',0,'myRIA','quickstart.js','startMyProgram']);

app.rq.push(['extension',1,'analytics_google','extensions/analytics_google.js','addTriggers']);
//app.rq.push(['extension',1,'bonding_buysafe','extensions/bonding_buysafe.js','addTriggers']);
//app.rq.push(['extension',1,'powerReviews','extensions/reviews_powerreviews.js','startExtension']);
//app.rq.push(['extension',0,'magicToolBox','extensions/imaging_magictoolbox.js','startExtension']); // (not working yet - ticket in to MTB)



//add tabs to product data.
app.rq.push(['templateFunction','productTemplate','onCompletes',function(P) {
	$( ".tabbedProductContent",$('#productTemplate_'+app.u.makeSafeHTMLId(P.pid))).tabs();
	}]);

app.rq.push(['script',0,(document.location.protocol == 'file:') ? app.vars.httpURL+'jquery/config.js' : app.vars.baseURL+'jquery/config.js']); //The config.js is dynamically generated.
app.rq.push(['script',0,app.vars.baseURL+'model.js']); //'validator':function(){return (typeof zoovyModel == 'function') ? true : false;}}
app.rq.push(['script',0,app.vars.baseURL+'includes.js']); //','validator':function(){return (typeof handlePogs == 'function') ? true : false;}})
app.rq.push(['script',1,app.vars.baseURL+'jeditable.js']); //used for making text editable (customer address). non-essential. loaded late.
app.rq.push(['script',0,app.vars.baseURL+'controller.js']);

//sample of an onDeparts. executed any time a user leaves this page/template type.
app.rq.push(['templateFunction','homepageTemplate','onDeparts',function(P) {app.u.dump("just left the homepage")}]);


//group any third party files together (regardless of pass) to make troubleshooting easier.
app.rq.push(['script',0,(document.location.protocol == 'https:' ? 'https:' : 'http:')+'//ajax.googleapis.com/ajax/libs/jqueryui/1.9.0/jquery-ui.js']);


/*
This function is overwritten once the controller is instantiated. 
Having a placeholder allows us to always reference the same messaging function, but not impede load time with a bulky error function.
*/
app.u.throwMessage = function(m)	{
	alert(m); 
	}

app.u.howManyPassZeroResourcesAreLoaded = function(debug)	{
	var L = app.vars.rq.length;
	var r = 0; //what is returned. total # of scripts that have finished loading.
	for(var i = 0; i < L; i++)	{
		if(app.vars.rq[i][app.vars.rq[i].length - 1] === true)	{
			r++;
			}
		if(debug)	{app.u.dump(" -> "+i+": "+app.vars.rq[i][2]+": "+app.vars.rq[i][app.vars.rq[i].length -1]);}
		}
	return r;
	}


//gets executed once controller.js is loaded.
//check dependencies and make sure all other .js files are done, then init controller.
//function will get re-executed if not all the scripts in app.vars.scripts pass 1 are done loading.
//the 'attempts' var is incremented each time the function is executed.

app.u.initMVC = function(attempts){
//	app.u.dump("app.u.initMVC activated ["+attempts+"]");
	var includesAreDone = true;

//what percentage of completion a single include represents (if 10 includes, each is 10%).
	var percentPerInclude = Math.round((100 / app.vars.rq.length));  
	var resourcesLoaded = app.u.howManyPassZeroResourcesAreLoaded();
	var percentComplete = resourcesLoaded * percentPerInclude; //used to sum how many includes have successfully loaded.

	$('#appPreViewProgressBar').val(percentComplete);
	$('#appPreViewProgressText').empty().append(percentComplete+"% Complete");

	if(resourcesLoaded == app.vars.rq.length)	{
//instantiate controller. handles all logic and communication between model and view.
//passing in app will extend app so all previously declared functions will exist in addition to all the built in functions.
//tmp is a throw away variable. app is what should be used as is referenced within the mvc.
		app.vars.rq = null; //to get here, all these resources have been loaded. nuke record to keep DOM clean and avoid any duplication.
		var tmp = new zController(app);
//instantiate wiki parser.
		myCreole = new Parse.Simple.Creole();
		}
	else if(attempts > 50)	{
		app.u.dump("WARNING! something went wrong in init.js");
		//this is 10 seconds of trying. something isn't going well.
		$('#appPreView').empty().append("<h2>Uh Oh. Something seems to have gone wrong. </h2><p>Several attempts were made to load the store but some necessary files were not found or could not load. We apologize for the inconvenience. Please try 'refresh' and see if that helps.<br><b>If the error persists, please contact the site administrator</b><br> - dev: see console.</p>");
		app.u.howManyPassZeroResourcesAreLoaded(true);
		}
	else	{
		setTimeout("app.u.initMVC("+(attempts+1)+")",250);
		}

	}



//Any code that needs to be executed after the app init has occured can go here.
//will pass in the page info object. (pageType, templateID, pid/navcat/show and more)
app.u.appInitComplete = function(P)	{
	app.u.dump("Executing myAppIsLoaded code...");
	}




//don't execute script till both jquery AND the dom are ready.
$(document).ready(function(){
	app.u.handleRQ(0)
	});






